OrbitQt/ConnectToStadiaWidget.cpp

```diff
@@ -0,0 +1,378 @@
+                               .arg(stadia_connection_->instance->display_name)
+                               .arg(QString::fromStdString(error.message())));
+      });
+
+  LOG("Deployment successful, grpc_port: %d", deployment_result.value().grpc_port);
+  CHECK(stadia_connection_->grpc_channel == nullptr);
+  std::string grpc_server_address =
+      absl::StrFormat("127.0.0.1:%d", deployment_result.value().grpc_port);
+  LOG("Starting gRPC channel to: %s", grpc_server_address);
+  stadia_connection_->grpc_channel = grpc::CreateCustomChannel(
+      grpc_server_address, grpc::InsecureChannelCredentials(), grpc::ChannelArguments());
+  CHECK(stadia_connection_->grpc_channel != nullptr);
+
+  emit Connected();
+}
+
+void ConnectToStadiaWidget::Disconnect() {
+  stadia_connection_->grpc_channel = nullptr;
+
+  // TODO(174561221) currently does not work
+  // if (stadia_connection_->service_deploy_manager != nullptr) {
+  //   stadia_connection_->service_deploy_manager->Shutdown();
+  // }
+  stadia_connection_->service_deploy_manager = nullptr;
+
+  ui_->rememberCheckBox->setChecked(false);
+
+  emit Disconnected();
+}
+
+void ConnectToStadiaWidget::OnConnectToStadiaRadioButtonClicked(bool checked) {
+  if (checked) {
+    emit Activated();
+  } else {
+    ui_->connectToStadiaInstanceRadioButton->setChecked(true);
+  }
+}
+
+void ConnectToStadiaWidget::OnErrorOccurred(const QString& message) {
+  if (isVisible()) {
+    QMessageBox::critical(this, QApplication::applicationName(), message);
+  } else {
+    ERROR("%s", message.toStdString());
+  }
+}
+
+void ConnectToStadiaWidget::OnSelectionChanged(const QModelIndex& current) {
+  if (!current.isValid()) return;
+
+  CHECK(current.model() == &instance_model_);
+  stadia_connection_->instance = current.data(Qt::UserRole).value<Instance>();
+  emit InstanceSelected();
+}
+
+void ConnectToStadiaWidget::OnRememberCheckBoxToggled(bool checked) {
+  QSettings settings;
+  if (checked) {
+    CHECK(stadia_connection_->instance != std::nullopt);
+    settings.setValue(kRememberChosenInstance, stadia_connection_->instance->id);
+  } else {
+    settings.remove(kRememberChosenInstance);
+  }
+}
+
+void ConnectToStadiaWidget::OnInstancesLoaded(
+    outcome::result<QVector<OrbitGgp::Instance>> instances) {
+  if (!instances) {
+    emit ErrorOccurred(QString("Orbit was unable to retrieve the list of available Stadia "
+                               "instances. The error message was: %1")
+                           .arg(QString::fromStdString(instances.error().message())));
+    return;
+  }
+
+  instance_model_.SetInstances(instances.value());
+  emit ReceivedInstances();
+
+  for (const auto& instance : instances.value()) {
+    std::string instance_id = instance.id.toStdString();
+
+    if (remembered_instance_id_ != std::nullopt) {
+      std::optional<int> row =
+          instance_model_.GetRowOfInstanceById(remembered_instance_id_.value());
```

Comment:
There is a generic way to make searches in a model: `QAbstractItemModel::match` (https://doc.qt.io/qt-5/qabstractitemmodel.html#match)

In that case you didn't need a new lookup function `GetRowOfInstanceById`. But I don't feel particurlarly strong about this.

Reply:
I like it and used it, thanks for the suggestion